### PR TITLE
Added training with custom data support, fixed bugs with ddpm

### PR DESCRIPTION
**Pull request by wu-ming233**
This pull request entails:
- Streamlining training on new data directly from midi files using an updated dataloader
- Supporting an arbitrary amount of tracks in the input, as opposed to the fixed (melody, bridge, piano) shape from POP909
- Extending the chord extractor with the option to round the duration of chords to the nearest beat. This significantly reduces error rate in preparing training data, while maintaining a reasonable accuracy
- Fixing `collate_fn` shifting unevenly in dataloader
- Fixing vanilla ddpm model not being able to train with `PolyffusionLearner`
- Fixing ddpm not being able to inference when `length=1`
- Enabling retraining from checkpoint with an increased `max_epochs`
- Implementing several convenient changes for batch inferencing

These changes are mainly just small additions to enhance user experience, and are completely backward-compatible.

### CODE REVIEW

File: polyffusion/data/dataloader.py [hunk 1]
```diff
@@ -19,7 +19,7 @@
 
 def collate_fn(batch, shift):
     def sample_shift():
-        return np.random.choice(np.arange(-6, 6), 1)[0]
+        return np.random.choice(np.arange(-6, 7), 1)[0]
```
Comment (aik2mlj) [inline polyffusion/data/dataloader.py]: This seems weird to me. We randomly transpose the key to 12 tones evenly. But now it is randomized to 13 tones. I think the former is correct.

File: polyffusion/main.py
Comment (aik2mlj) [inline polyffusion/main.py:1]: Let's add `data_dir=args.data_dir` to all the model configs!

File: polyffusion/prepare_data.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+def force_length(music : muspy.music, bars = 8):
+    for track in music.tracks:
+        timesteps = track.get_end_time()
+        old_bars = (timesteps + 15) // 16
+        div = bars // old_bars
+        for i in range(1, div):
+            tmp = track.deepcopy()
+            tmp.adjust_time(lambda x : x + i * timesteps)
+            track.notes.extend(tmp.notes)
```
Comment (aik2mlj) [inline polyffusion/prepare_data.py]: This function seems weird. The last `extend` function duplicates the notes many times.

File: polyffusion/train/__init__.py [hunk 3]
```diff
@@ -23,11 +23,13 @@ def __init__(self, params, param_scheduler, output_dir) -> None:
...
-                if params != self.params:
-                    print("New params differ, using former params instead...")
-                    params = self.params
+            with open(f"{output_dir}/params.json", "r+") as params_file:
+                old_params = AttrDict(json.load(params_file))
+                if old_params != self.params:
+                    print("New params differ, using new params could break things")
+                    params_file.seek(0)
+                    json.dump(self.params, params_file)
+                    params_file.truncate()
```
Comment (aik2mlj) [inline polyffusion/train/__init__.py]: Maybe we can leave the old param file unchange since it records the params when training, which should be preserved.

File: polyffusion/train/__init__.py [hunk 4]
```diff
@@ -40,7 +42,6 @@ def train(self):
             if input("Resume training? (y/n)") != "y":
                 return
         else:
-            output_dir = f"{output_dir}/{datetime.now().strftime('%m-%d_%H%M%S')}"
```
Comment (aik2mlj) [inline polyffusion/train/__init__.py]: This line is intended. We want the `output_dir` to be able to host multiple trainings with different time stamps.

**Review by aik2mlj — CHANGES_REQUESTED** [review]
Thanks for your great work! I have some concerns as listed below, could you please take your time to review them?

File: polyffusion/data/dataloader.py (hunk 1 above)
Comment (wu-ming233) [reply polyffusion/data/dataloader.py]: In numpy's documentation, `np.arange(start, stop)` generates within the half-open interval `[start, stop)`. Thus, to generate 12 tones evenly, I believe we would need (-6, 7). See https://numpy.org/doc/stable/reference/generated/numpy.arange.html

File: polyffusion/main.py
Comment (wu-ming233) [reply polyffusion/main.py:1]: Agreed!

File: polyffusion/prepare_data.py (hunk 2 above)
Comment (wu-ming233) [reply polyffusion/prepare_data.py]: Sorry for the lack of documentation. This function is for the case when the user wants to train short loops which are usually either 4-bars or 8-bars in length per MIDI. For the ones that are 4-bars, we use this function to loop it twice. Note that this `force_length` function is disabled by default when preparing data.

It is true that this would not function as intended if a MIDI has 1 track that's >= 8 bars but some other tracks < 8 bars, as it would loop the tracks < 8 bars. To address this, I implemented a simple check in the newest commit.

File: polyffusion/train/__init__.py (hunk 3 above)
Comment (wu-ming233) [reply polyffusion/train/__init__.py]: Agreed! I implemented that, and allowed users to choose if they want to overwrite the old params file or to save the old params file with a time stamp.

File: polyffusion/train/__init__.py (hunk 4 above)
Comment (wu-ming233) [reply polyffusion/train/__init__.py]: That makes sense. I have reverted the change.

**Comment by wu-ming233** [comment]
@aik2mlj The new commit implements the changes requested above, as well as two minor things:
1. When preparing data from MIDI, we would like to discard the non-melody parts, including drum tracks and melody tracks that have instrument program number >= 113. I've added this option in `prepare_data.py`, and it's enabled by default.
2. Unconditional vanilla ddpm lacked a way to specify which weights to use under the same checkpoint. I've added that, in a similar fashion as what's done in the other ldm models.

Let me know if anything else needs to be changed!

File: polyffusion/data/dataloader.py (hunk 1 above)
Comment (aik2mlj) [reply polyffusion/data/dataloader.py]: Well I know, but [-6, 6) is transposing 12 tones evenly. If it's [-6, 7), the -6 transposed and the +6 transposed versions are on the same key.

**Comment by aik2mlj** [comment]
Looks good to me! Thank you for your efforts.
Now it's only the augmentation range [-6, 7) that needs to be fixed.

File: polyffusion/data/dataloader.py (hunk 1 above)
Comment (wu-ming233) [reply polyffusion/data/dataloader.py]: My mistake. Sorry about that! Commit added
